lib/Serialization/Deserialization.cpp

```diff
@@ -166,10 +167,27 @@ static void skipRecord(llvm::BitstreamCursor &cursor, unsigned recordKind) {
 }
 
 void ModuleFile::fatal(llvm::Error error) const {
+  Core->fatal(diagnoseFatal(std::move(error)));
+}
+
+llvm::Error ModuleFile::diagnoseFatal(llvm::Error error) const {
   if (FileContext)
     getContext().Diags.diagnose(SourceLoc(), diag::serialization_fatal,
                                 Core->Name);
-  Core->fatal(std::move(error));
+  std::string msg;
+  {
+    llvm::raw_string_ostream os(msg);
+    Core->outputDiagnosticInfo(os);
+  }
+  msg += ": " + toString(std::move(error));
+  llvm::dbgs() << "*** DESERIALIZATION FAILURE ***\n";
+  llvm::dbgs() << "*** If any module named here was modified in the SDK, please delete the ***\n";
+  llvm::dbgs() << "*** new swiftmodule files from the SDK and keep only swiftinterfaces.   ***\n";
+  outputDiagnosticInfo(llvm::dbgs());
+  llvm::dbgs() << '\n' << msg << '\n';
+  assert(false && "fatal deserialization failure");
```

Comment:
Limiting the crash option to debug builds is not practical. Since we have a precise serialization format we need to use the same compiler that built the SDK to investigate most deserialization issues. Getting the precise errors in debug only would mean we'd have to rebuild the specific compiler tag on every deserialization failure we investigate. That's not very sustainable in term on engineer-hour per bug.

How about we limit the new behavior for LLDB only and always keep the hard crash for the compiler?

Reply:
Done!